lib/*fs, lib/vfscore: Add proper `dirent/dirent64` structure usage

StefanJum opened the pull request:
Unikraft defines the `dirent` structure as identical to `dirent64`.
This works fine if not using Unikraft in binary compatibility mode,
since we have total control over the internal functions/structures that
are used.

However, if we use Unikraft in binary compatibility mode, we want to use
Linux `dirent` and `dirent64` structures, in order to maintain compatibility
with older libc versions that do not use the `*64` calls by default.
Since the filesystems `READDIR` vop uses `dirent64`, we will convert to a
`dirent` structure within the `readdir_r` call.

You can see the issue happening in #919 
In order to test it with the `app-elfloader`, you can use the `ls` executable from the [`dynamic-apps` repository](https://github.com/unikraft/dynamic-apps/tree/master/ls), which uses an old libc version that calls `getdents` instead of `getdents64`.

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.

StefanJum committed:
lib/vfscore: Add non-largefile variant for dirent

Unikraft defines the `dirent` structure as identical to `dirent64`.
This works fine if not using Unikraft in binary compatibility mode,
since we have total control over the internal functions/structures that
are used.

However, if we use Unikraft in binary compatibility mode, we want to use
Linux `dirent` and `dirent64` structures in order to maintain compatibility
with older libc versions that do not use the `*64` calls by default.
Since the filesystems `READDIR` vop uses `dirent64`, we will convert to a
`dirent` structure within the `readdir_r` call.

Signed-off-by: Stefan Jumarea <stefanjumarea02@gmail.com>
GitHub-Closes: #919

skuenzer commented on Config.uk:
```diff
@@ -180,6 +180,13 @@ if RECORD_BUILDTIME
 	endchoice
 endif
 
+config BINARY_COMPAT
```
Can't we provide an option with `vfscore` instead? This has to do with largefile support only (so a VFS thing), so I'd say to add an option there instead that enables the `64` bits or `32` bits version of the system calls (and also registering those variants with `UK_LLSYSCALL_R_DEFINE()`). The key would be to keep the behavior as is when we natively compile (e.g., via preprocessor renaming tricks as of today) and only get either the old style or new style system call handlers registered when the option is off. With `app-elfloader` we would enable (`imply`) this option in vfscore for binary compatibility. Alternatively, the default of this option in vfscore could be `y` if binary system calls are enabled in syscall_shim:
```KConfig
default n
default y if `LIBSYSCALL_SHIM_HANDLER`
```

mogasergiu commented:
Just a more general comment, but would it be a good idea to use `unlikely`'s on error path's here as well? :thinking:

StefanJum commented:
Done @skuenzer, I moved the config option in vfscore, it will get enabled by default if the syscall handler is enabled.

Also added the `unlikely`'s as @mogasergiu suggested .

RaduNichita commented on lib/nolibc/include/dirent.h:
```diff
@@ -19,21 +19,40 @@ typedef struct __dirstream DIR;
 };
 
+#ifdef CONFIG_NONLARGEFILE_SYSCALLS
+struct dirent {
```
Could we keep the order of members in this struct as in the `dirent` struct?

StefanJum replied:
No. It's supposed to work in binary compatibility mode, the order should be the one the libc uses (referenced in [the man pages](https://man7.org/linux/man-pages/man2/getdents.2.html)).

skuenzer replied:
This is right. Maybe add a comment to the struct definitions that they have to match Linux system call ABI.

RaduNichita commented on lib/vfscore/main.c:
```diff
@@ -1167,36 +1182,85 @@ int readdir_r(DIR *dir, struct dirent *entry, struct dirent **result)
 	struct vfscore_file *fp;
 
+	/**
+	 * `sys_readdir` will end up calling `VOP_READDIR`, which expects
```
```suggestion
	 * `sys_readdir()` will end up calling `VOP_READDIR`, which expects
```

RaduNichita reviewed (changes requested):
Thanks a lot for providing the fix, @StefanJum. I have provided some nitpicks. 

I will approve the PR after that! :muscle:

StefanJum commented:
Done @RaduNichita, thank you.

RaduNichita reviewed (approved):
All good now. Thanks, @StefanJum! :100: 

Reviewed-by: Radu Nichita <radunichita99@gmail.com>

mogasergiu commented on lib/vfscore/main.c:
```diff
@@ -1167,36 +1182,85 @@ int readdir_r(DIR *dir, struct dirent *entry, struct dirent **result)
 	}
 	// Our dirent has (like Linux) a d_reclen field, but a constant size.
 	entry->d_reclen = sizeof(*entry);
+	entry->d_ino = entry64.d_ino;
```
Is this really fine :thinking:? You are referencing the fields of `entry64`, which may have not been initialized unless we entered the `if` statement block above, shouldn't you probably do this in the if block statement above :thinking:? Also this whole function looks like it's asking to be re-ordered. that `if` statement block looks... unnecessarily imbricated/large.

mogasergiu commented on lib/vfscore/main.c:
```diff
@@ -1167,36 +1182,85 @@ int readdir_r(DIR *dir, struct dirent *entry, struct dirent **result)
+	 * `sys_readdir()` will end up calling `VOP_READDIR`, which expects
+	 * a dirent64 structure instead of the `struct dirent` entry parameter.
+	 */
+	struct dirent64 entry64;
```
While we're still at it, have declarations grouped in a single block, in a reverse xmas tree manner, i.e.:
```C
        /**
	 * `sys_readdir()` will end up calling `VOP_READDIR`, which expects
	 * a dirent64 structure instead of the `struct dirent` entry parameter.
	 */
        struct dirent64 entry64;
	struct vfscore_file *fp;
	int error;
```

mogasergiu commented on lib/vfscore/main.c:
```diff
@@ -1167,36 +1182,85 @@ int readdir_r(DIR *dir, struct dirent *entry, struct dirent **result)
-		__attribute__((alias("readdir_r")));
+{
+	int error;
+	struct vfscore_file *fp;
```
Reverse xmas tree :parrot:.

mogasergiu reviewed (changes requested):
Looks cool to me, some comments for your notification box :mailbox:.

StefanJum commented:
Done @mogasergiu, thank you

mogasergiu commented on lib/vfscore/main.c:
```diff
@@ -1164,39 +1179,93 @@ struct dirent *readdir(DIR *dir)
+
+	error = sys_readdir(fp, &entry64);
+	fdrop(fp);
+	if (unlikely(error)) {
```
Do we really need three `unlikely`s for same `error` :thinking:? And those `else` statements look somewhat unnecessary on the surface :thinking:. What is your opinion on this alternative:
```C
int readdir_r(DIR *dir, struct dirent *entry, struct dirent **result)
{
	/**
	 * `sys_readdir()` will end up calling `VOP_READDIR`, which expects
	 * a dirent64 structure instead of the `struct dirent` entry parameter.
	 */
	struct dirent64 entry64;
	struct vfscore_file *fp;
	int error;

	*result = NULL;

	// Our dirent has (like Linux) a d_reclen field, but a constant size.
	entry->d_reclen = sizeof(*entry);

	trace_vfs_readdir(dir->fd, entry);
	error = fget(dir->fd, &fp);
	if (unlikely(error))
		return error;    // This here can never be `ENOENT`, so goto err_out check is unnecessary

	error = sys_readdir(fp, &entry64);
	fdrop(fp);
	if (unlikely(error)) {
		trace_vfs_readdir_err(error);

		return error == ENOENT ? 0 : error;
	}

	trace_vfs_readdir_ret();

	entry->d_ino = entry64.d_ino;
	entry->d_type = entry64.d_type;
	entry->d_off = entry64.d_off;
	memcpy(entry->d_name, entry64.d_name, sizeof(entry64.d_name));

       *result = entry;

	return 0;
}
```

mogasergiu commented on lib/vfscore/main.c:
```diff
@@ -1164,39 +1179,93 @@ struct dirent *readdir(DIR *dir)
+struct dirent64 *readdir64(DIR *dir)
+{
+	static __thread struct dirent64 entry, *result;
+	int ret;
```
Same here. What do you think of this :thinking::
```C
struct dirent64 *readdir64(DIR *dir)
{
	static __thread struct dirent64 entry, *result;

        /* By definition, `readdir64_r` already places NULL into `result` if end of directory reached */
	errno = readdir64_r(dir, &entry, &result);

	return result;
}
```

mogasergiu replied:
Oh, one more thing, does `result` really need to be in the TLS here, I am not really sure of that? There is already `entry` keeping track of this, is that not enough :thinking:.

StefanJum replied:
Fair, I don't think that's needed there. I just copy pasted the implementation we have for [`readdir`](https://github.com/unikraft/unikraft/blob/staging/lib/vfscore/main.c#L1150), so no mind filter involved :laughing: 
I'll remove `result` from the TLS.

mogasergiu replied:
I know what that's like :rofl:. So `readdir` is flawed too :laughing:? This is going to look very inconsistent then, perhaps you should add an additional commit changing that? So, first commit of the PR the one that changes the existing `readdir` removing `result` from TLS and second commit of the PR the one you have right now :1234:. What do you think?

StefanJum replied:
Yes, I'll do that :laughing:

mogasergiu reviewed (changes requested):
Awesome, I appreciate the very fast response :smile:. I may have a few more comments. Let me know what you think :thinking:.

StefanJum committed:
lib/vfscore: Remove readdir_r result from TLS

The `result` structure of the `readdir_r` function will not need to be
part of the TLS, the `entry` structure alone is enough.

Signed-off-by: Stefan Jumarea <stefanjumarea02@gmail.com>

StefanJum commented:
Done @mogasergiu, I updated the `readdir64_r` too.

mogasergiu reviewed (approved):
That was quick! Thank you!

Reviewed-by: Sergiu Moga <sergiu@unikraft.io>

skuenzer commented on lib/vfscore/Config.uk:
```diff
@@ -25,6 +25,16 @@ help
 	be changed with the following library parameters:
 	'vfs.rootfs', 'vfs.rootdev', 'vfs.rootflags', and 'vfs.rootopts'
 
+config NONLARGEFILE_SYSCALLS
```
Please namespace the config option, like:
`LIBVFSCORE_NONLARGEFILE`

skuenzer commented on lib/vfscore/Config.uk:
```diff
@@ -25,6 +25,16 @@ help
 	'vfs.rootfs', 'vfs.rootdev', 'vfs.rootflags', and 'vfs.rootopts'
 
+config NONLARGEFILE_SYSCALLS
+	bool "Provide proper legacy non largefile syscalls"
```
I would shorten the config title here to
`"Non-largefile system calls"`

skuenzer commented on lib/vfscore/Config.uk:
```diff
@@ -25,6 +25,16 @@ help
+		Add different dirent and dirent64 structure and their respective
+		syscalls (i.e. treat Linux non largefile legacy syscalls properly).
+		If this option is not set, the dirent structure will be aliased to
+		dirent64, so will all other f/f64 functions.
```
I'd mention that only the 64-bit variants of system calls are available via the syscall_shim, in case the option is off. Something like:
```diff
-dirent64, so will all other f/f64 functions.
+dirent64, so will all other f/f64 functions. If lib/syscall_shim
+is enabled, only the 64-bit version of system calls are registered.
```

skuenzer replied:
What does `f/f64` exactly mean? I can guess it.

StefanJum replied:
> What does `f/f64` exactly mean? I can guess it.

The non-largefile/largefile function pairs, I rephrased it, it was unclear.

skuenzer commented on lib/vfscore/Config.uk:
```diff
@@ -25,6 +25,16 @@ help
 
+config NONLARGEFILE_SYSCALLS
+	bool "Provide proper legacy non largefile syscalls"
+	default y if LIBSYSCALL_SHIM_HANDLER
```
👍🏻

skuenzer commented on lib/nolibc/include/dirent.h:
```diff
@@ -57,17 +76,16 @@ long           telldir(DIR *);
-#define readdir64 readdir
-#define readdir64_r readdir_r
+
 #define scandir64 scandir
```
Do these functions also need special handling in the context of `dirent`? If yes, can you provide it or at least leave a `TODO` comment? If not, can you add a comment why those do not need to be handled special?

StefanJum replied:
The `scandir64` I just missed. I provided it now.

The `alphasort` and `versionsort` are not provided by the Unikraft core, so we can just let the defines there. I added a comment about that in the file.

skuenzer reviewed (changes requested):
Hi @StefanJum, thanks a lot for this work. I have a few comments in the code and some general queastions:
- Do you think it would make sense that with this PR you only register non-largefile system calls to syscall_shim if `CONFIG_NONLARGEFILE_SYSCALLS` is set? This would means that you remove the corresponding `UK_SYSCALL_DEFINE` and `UK_PROVIDED_SYSCALLS` registrations depending on this option.
- Can we change the commit title from
`lib/*fs, lib/vfscore: Add proper dirent structure` to
`lib/vfscore: Non-largefile variant for dirent`. Because the updates of the filesystems are an implication because of the changes in lib/vfscore, I think not explicitly mentioning them is fine.
- This PR is only covering `dirent`/`dirent64`, which is fine, I just wonder what we do with `read`, `read64` and company.
- A minor thing: If you check for your option to be set, you can use `#if CONFIG_ABC` instead of `#ifdef CONFIG_ABC`. This is slightly more powerful if you manually do `#define CONFIG_ABC 0`.

StefanJum commented:
> Hi @StefanJum, thanks a lot for this work. I have a few comments in the code and some general queastions:
> 
>     * Do you think it would make sense that with this PR you only register non-largefile system calls to syscall_shim if `CONFIG_NONLARGEFILE_SYSCALLS` is set? This would means that you remove the corresponding `UK_SYSCALL_DEFINE` and `UK_PROVIDED_SYSCALLS` registrations depending on this option.

done

>     * Can we change the commit title from
>       `lib/*fs, lib/vfscore: Add proper dirent structure` to
>       `lib/vfscore: Non-largefile variant for dirent`. Because the updates of the filesystems are an implication because of the changes in lib/vfscore, I think not explicitly mentioning them is fine.

done

>     * This PR is only covering `dirent`/`dirent64`, which is fine, I just wonder what we do with `read`, `read64` and company.

Yes, we also need similar updates for any other structure that is different (e.g. `statfs64`), this covers only the `getdeents`, `readdir/_r` and `scandir` functions (the ones that use the `dirent/64` structure). I think we can open an issue on full non-largefile support so we have it in mind.

>     * A minor thing: If you check for your option to be set, you can use `#if CONFIG_ABC` instead of `#ifdef CONFIG_ABC`. This is slightly more powerful if you manually do `#define CONFIG_ABC 0`.

done

StefanJum commented:
All done @skuenzer, sorry for the force-push spam, there were some checkpatch and ci/cd issues.

StefanJum commented:
> WARNING:LINE_SPACING: Missing a blank line after declarations
#233: FILE: lib/vfscore/main.c:1080:
> 	int (*cmp)(const struct dirent **, const struct dirent **))
> 	\_\_attribute\_\_((alias("scandir64")));

The checkpatch is getting confused by the function pointer argument here :laughing:

mogasergiu commented:
> > WARNING:LINE_SPACING: Missing a blank line after declarations
> > #233: FILE: lib/vfscore/main.c:1080:
> > int (*cmp)(const struct dirent **, const struct dirent **))
> > __attribute__((alias("scandir64")));
> 
> The checkpatch is getting confused by the function pointer argument here laughing

The whole `scandir64` function looks like a big checkpatch violation to be honest :triumph:. Maybe we could rewrite it better :thinking:.

StefanJum commented:
> The whole `scandir64` function looks like a big checkpatch violation to be honest 😤. Maybe we could rewrite it better 🤔.

Yes, it's absolutely hideous, I just copied the `scandir` function. I'll try to do a rewrite if I catch some time.

skuenzer commented:
> ⚠️ **Checkpatch passed with warnings**
> 
> Beep boop! I ran Unikraft's [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) support script on your pull request but it encountered warnings:
> SHA 	commit 	checkpatch
> [cfc5c4c](https://github.com/unikraft/unikraft/commit/cfc5c4c73a952eed1bfdc7f335cad3e55c387f36) 	lib/vfscore: Remove readdir_r result from TLS 	✅
> [093be54](https://github.com/unikraft/unikraft/commit/093be545512032212ca524e56692bd60b40d087d) 	lib/vfscore: Add non-largefile variant for dirent 	⚠️
> 
> **Truncated logs starting from first warning [093be54](https://github.com/unikraft/unikraft/commit/093be545512032212ca524e56692bd60b40d087d)**:
> 
> ```
> WARNING:LONG_LINE: line over 80 characters
> #68: FILE: lib/nolibc/include/dirent.h:22:
> +#if !defined(CONFIG_LIBVFSCORE_NONLARGEFILE) || (CONFIG_LIBVFSCORE_NONLARGEFILE <= 0)
> 
> WARNING:LINE_SPACING: Missing a blank line after declarations
> #233: FILE: lib/vfscore/main.c:1080:
> +	int (*cmp)(const struct dirent **, const struct dirent **))
> +	__attribute__((alias("scandir64")));
> 
> total: 0 errors, 2 warnings, 377 lines checked
> 
> NOTE: For some of the reported defects, checkpatch may be able to
>       mechanically convert to the typical style using --fix or --fix-inplace.
> 
> /tmp/build/a53d4c5b/patches/0002-lib-vfscore-Add-non-largefile-variant-for-dirent.patch has style problems, please review.
> 
> NOTE: Ignored message types: ASSIGN_IN_IF FILE_PATH_CHANGES NEW_TYPEDEFS OBSOLETE
> 
> NOTE: If any of the errors are false positives, please report
>       them to the maintainer, see CHECKPATCH in MAINTAINERS.
> ```
> 
> [View complete logs](https://builds.unikraft.io/builds/19973093) | Learn more about [Unikraft's coding style](https://github.com/unikraft/unikraft/blob/staging/CODING_STYLE.md) and [contribution guidelines](https://github.com/unikraft/unikraft/blob/staging/CONTRIBUTING.md).

Because I am seeing this here:
You do not need to do
```
#if !defined(CONFIG_LIBVFSCORE_NONLARGEFILE) || (CONFIG_LIBVFSCORE_NONLARGEFILE <= 0)
```
the following achieves exactly the same (except it does not check for < 0, but it is fine to assume a CONFIG to be set to 0(and undefined) or 1):
```
+#if !CONFIG_LIBVFSCORE_NONLARGEFILE
```
The preprocessor handles the case of being not defined correctly.

StefanJum commented:
> The preprocessor handles the case of being not defined correctly.

I did not know that :eyes: Pushed the change.

skuenzer commented on lib/nolibc/include/dirent.h:
```diff
@@ -19,21 +19,44 @@ typedef struct __dirstream DIR;
-struct dirent {
+#if !CONFIG_LIBVFSCORE_NONLARGEFILE
+#define dirent dirent64
+#endif /* CONFIG_LIBVFSCORE_NONLARGEFILE */
```
```diff
-#endif /* CONFIG_LIBVFSCORE_NONLARGEFILE */
+#endif /* !CONFIG_LIBVFSCORE_NONLARGEFILE */
```

skuenzer replied:
but maybe it is better to move this alias definition as `#else` under `struct dirent` (line 47)

skuenzer commented on lib/vfscore/main.c:
```diff
@@ -1068,8 +1073,67 @@ int scandir(const char *path, struct dirent ***res,
 	*res = names;
 	return cnt;
 }
+#else
```
```diff
-#else
+#else /* !CONFIG_LIBVFSCORE_NONLARGEFILE */
```

skuenzer replied:
Reason is actually what we did so far in the code (needs to be inlcuded in our convention):
```c
#if CONFIG_ABC
...
#else /* !CONFIG_ABC */ <--- negation here because it is like "#elif !CONFIG_ABC"
...
#endif /* !CONFIG_ABC */ <--- repeats the last condition we had
```

skuenzer commented on lib/vfscore/main.c:
```diff
@@ -1068,8 +1073,67 @@ int scandir(const char *path, struct dirent ***res,
+	int (*sel)(const struct dirent *),
+	int (*cmp)(const struct dirent **, const struct dirent **))
+	__attribute__((alias("scandir64")));
+#endif /* CONFIG_LIBVFSCORE_NONLARGEFILE */
```
```diff
-#endif /* CONFIG_LIBVFSCORE_NONLARGEFILE */
+#endif /* !CONFIG_LIBVFSCORE_NONLARGEFILE */
```

skuenzer commented on lib/vfscore/main.c:
```diff
@@ -1158,39 +1234,91 @@ struct dirent *readdir(DIR *dir)
+
+	return 0;
 }
+#else
```
```c
#else /* !CONFIG_LIBVFSCORE_NONLARGEFILE */
```

skuenzer commented on lib/vfscore/main.c:
```diff
@@ -1158,39 +1234,91 @@ struct dirent *readdir(DIR *dir)
+struct dirent *readdir(DIR *dir) __attribute__((alias("readdir64")));
+int readdir_r(DIR *dir, struct dirent *entry, struct dirent **result)
+	 __attribute__((alias("readdir64_r")));
+#endif /* CONFIG_LIBVFSCORE_NONLARGEFILE */
```
```c
#endif /* !CONFIG_LIBVFSCORE_NONLARGEFILE */
```

skuenzer commented on lib/vfscore/main.c:
```diff
@@ -1158,39 +1234,91 @@ struct dirent *readdir(DIR *dir)
+#endif /* CONFIG_LIBVFSCORE_NONLARGEFILE */
 
-// FIXME: in 64bit dirent64 and dirent are identical, so it's safe to alias
 #undef readdir64_r
```
Do we need to guard this `#undef` with `#ifdef`?

StefanJum replied:
Yes, it was already there so I missed it.

StefanJum commented:
Updated @skuenzer, thanks.

skuenzer commented on lib/vfscore/main.c:
```diff
@@ -1158,39 +1234,93 @@ struct dirent *readdir(DIR *dir)
+	return 0;
+}
 
 #undef readdir64
```
Same would be here then (guarding).

StefanJum replied:
Solved @skuenzer

skuenzer commented on lib/vfscore/Config.uk:
```diff
@@ -22,6 +22,19 @@ help
 	Automatically mounts '/' during boot. The parameters are
 	compiled-in and cannot be changed at runtime.
 
+config LIBVFSCORE_NONLARGEFILE
```
Sorry, I did not see this, can you move this option before `LIBVFSCORE_AUTOMOUNT_ROOTFS`? It is now splitting up the connected block of `LIBVFSCORE_AUTOMOUNT_ROOTFS` and `LIBVFSCORE_ROOTFS`, etc.

StefanJum replied:
Ah, yes, I missed that. Fixed.

skuenzer reviewed (approved):
Thanks a lot for your work!

Reviewed-by: Simon Kuenzer <simon@unikraft.io>
Approved-by: Simon Kuenzer <simon@unikraft.io>
